First Implementaion of ilAtomQuery with an Example in ilForumTopic

chfsx opened the pull request.

chfsx committed:
First Implementaion of ilAtomQuery with an Example in ilForumTopic

chfsx commented on Modules/Forum/classes/class.ilForumTopic.php:
```diff
@@ -565,40 +565,38 @@ public function movePosts($old_obj_id, $old_pk, $new_obj_id, $new_pk)
-			);
-
-			$this->db->manipulateF('
+			$ilAtomQuery = new ilAtomQuery($ilDB);
```
This is an example how to replace current table locks with an abstraction for galera support

mjansenDatabay replied:
Well done @chfsx, looks easy to use. I also checked the code in **\ilBuddySystemRelationRepository::save** which could be easily changed as well.

@bheyser: This approach is (of course) a little different compared to your **\ilTestProcessLocker**. I only found two usages of the respective methods. In my opinion your lock mechanism has to require a closure (instead of calling lock/unlock in the consuming code), and in case of database locking (set in the assessment administration) it should pass this closure to **\ilAtomQuery**.

chfsx committed:
query-closures have do be done one after another, therefore ilAtomQuery has to allow multiple queries

chfsx committed:
separated lock-handling

chfsx committed:
comments and descriptions

chfsx committed:
ilAtomQuery: split queries in multiple closures

chfsx commented on Modules/Forum/classes/class.ilForumTopic.php:
```diff
@@ -565,40 +565,43 @@ public function movePosts($old_obj_id, $old_pk, $new_obj_id, $new_pk)
+					array('integer', 'integer'),
+					array($new_obj_id, $current_id));
+			});
+			$ilAtomQuery->addQueryClosure(function ($ilDB) use ($new_obj_id, $current_id) {
```
I splitted the queries into multiple Closures which should allow to rerun only to not failed queries (will follow in the next commit)

chfsx commented on Services/Database/classes/Atom/class.ilAtomQuery.php:
```diff
@@ -0,0 +1,320 @@
+				$this->ilDBInstance->commit();
+			} catch (ilDatabaseException $e) {
+				if ($i > 10) {
+					throw $e;
```
Is there a better way to prevent ilAtomQuery from beeing stuck in a endless loop?

mjansenDatabay replied:
Breaking after a certain number of attempts was the first approach that came into my mind when I read this last weekend.

alex40724 replied:
[deleted] Stefan gave me already the answers... Maybe waiting some random time between the tries could help, but I am lacking experience with this.

chfsx replied:
The last commits optimized this and tests the break after 10 ilDatabaseException

chfsx commented on Services/Database/test/Atom/ilDatabaseAtomBaseTest.php:
```diff
@@ -0,0 +1,86 @@
+ * @author  Fabian Schmid <fs@studer-raimann.ch>
+ * @version 1.0.0
+ */
+class ilDatabaseAtomBaseTest extends PHPUnit_Framework_TestCase {
```
Tests on database will follow here...

chfsx committed:
deprecated lockTables and unlockTables

mjansenDatabay commented on Services/Database/classes/Atom/class.ilAtomQuery.php:
```diff
@@ -0,0 +1,320 @@
+	 *
+	 * @param \Closure $query
+	 */
+	public function addQueryClosure(Closure $query) {
```
I suggest to change the expected type to _Callable_. This allows passing an object by implementing the **__invoke()** method.

klees replied:
I would prefer if we would stick to the `Closure`. `Callable` introduces all kinds of weirdness due to strings or arrays interpreted as names of some function or method. IMO that's legacy from PHPs wild teenage years and will be removed some times soon, as PHP is becoming a grown up now. If anyone really needs it, one could wrap the callable in a closure and use the API anyway. The weirdness is then at least contained at the side of the caller.

chfsx committed:
added new DB-Type Galera

chfsx committed:
switched to callable or queries

chfsx committed:
track running query

chfsx committed:
add write lock directly

chfsx committed:
improved tests

chfsx committed:
split tests

chfsx committed:
switched tests to callable and implemented more tests, especially handling multiple exceptions

chfsx committed:
more precice PHPDoc for passing callables to ilAtomQuery

klees commented on Modules/Forum/classes/class.ilForumTopic.php:
```diff
@@ -565,40 +565,43 @@ public function movePosts($old_obj_id, $old_pk, $new_obj_id, $new_pk)
+			$current_id = $this->id;
 
-			$this->db->manipulateF('
+			$ilAtomQuery = new ilAtomQuery($ilDB);
```
Why not $ilDB->buildAtomQuery()? You could force the users of AtomQuery to code against an interface then, instead if requiring them to use a certain implementation. This will help one the AtomQuery gets more functionality which need to be implemented differently for different DBs.

klees commented on Services/Database/classes/Atom/class.ilAtomQuery.php:
```diff
@@ -0,0 +1,320 @@
+			throw new ilDatabaseException('The selected Isolation-level is not allowd when locking tables with write-locks');
+		}
+
+		if ($this->ilDBInstance->supportsTransactions()) {
```
This would not be necessary if the database would construct the ilAtomQuery (instead of the user), see above.

chfsx replied:
The current version is based on ilDBInterface::runAtomQuery(); The points I  don't like in this version is the fact that you have to know the structure of the array for the locks (as in the old implementation) and you can't seperate queries to multiple callables which would allow to run them only once when one is failing.

Both of those points could be optimized by provide method chaining...

klees commented on Services/Database/classes/Atom/class.ilAtomQuery.php:
```diff
@@ -0,0 +1,320 @@
+	 * Add table-names which are influenced by your queries, MyISAm has to lock those tables. Lock
+	 *
+	 * @param string $table_name
+	 * @param int $lock_level use ilAtomQuery::LOCK_READ or ilAtomQuery::LOCK_WRITE
```
How is the dependency between LOCK_LEVELS and ISOLATION_LEVELS? I.e., what does it mean when i use e.g. ISOLATION_READ_UNCOMMITED with LOCK_WRITE? How is the expected behaviour of the DB?

chfsx replied:
currently only ISOLATION_SERIALIZABLE is possible, which is also needed when firing LOCK_WRITEs

chfsx replied:
Maybe it'll be better to switch that and only choose the ISOLATION_LEVEL which sets the locking for the tables. I'll switch that for the next push

klees commented on Services/Database/classes/Atom/class.ilAtomQuery.php:
```diff
@@ -0,0 +1,320 @@
+	 * @throws \ilDatabaseException
+	 */
+	public function run() {
+		self::checkIsolationLevel($this->getIsolationLevel());
```
Why not check the isolation on setIsolationLevel/in the constructor?

chfsx committed:
implemented getSequenceName on ilDB

chfsx committed:
check isolation level during construct

chfsx committed:
possibility to lock sequences, too

chfsx committed:
test with locking tables

chfsx committed:
check table-name

chfsx committed:
ilAtomQuery on ilDB

chfsx committed:
use ilDBInterface for AtomQuery

chfsx committed:
setuup table for every test

chfsx committed:
add second galera instance

chfsx committed:
renamed test

chfsx committed:
replaced lock infos and callables with standard

chfsx committed:
check against values in DB

chfsx committed:
splitted ilAtomQuery to ilAtomQueryTransaction and ilAtomQueryLock
ilDBInterface::buildAtomQuery() return the implementation which is supported by the DatabaseEngine

chfsx committed:
run all queries in one block

chfsx committed:
missing require and correct ilAtom child

chfsx committed:
add Atom-tests to Database Testsuite

chfsx committed:
switched ilForumTopic to new implementation

chfsx committed:
improved callable-checks

chfsx committed:
check all types of allowed closures

chfsx committed:
do not set the table locking on your own, let ilAtomQuery determine the locklevel (currently only WRITE supported, which is the only used in ILIAS)

chfsx committed:
allow ilAtomQueries without tables

chfsx committed:
switched to new ilAtomQueryException

chfsx committed:
allow only one Query-Closure per run

chfsx committed:
switched to other class in test to prevent warnings

chfsx committed:
improved callable check, test and implementation in ilForumTopic

chfsx committed:
only allow one parameter in closure and this must be typehinted with ilDBInstance

chfsx commented:
@alex40724 ilAtomQuery->checkCallable() now checks if the Closure has hinted the first (and only) parameter with ilDBInterface

chfsx committed:
removed unsed test

chfsx committed:
ilDBInterface: locks as direct exec-statement on PDO

chfsx committed:
fixed tests for locks with tables
